### PR TITLE
Improve TypeScript definition

### PR DESCRIPTION
**Pull request by Richienb**
If an array is provided to `.all`, a more precise type definition is used with a fallback for when iterables are used. `.fn` also offers more precise typings for the returned function.

Related to https://github.com/sindresorhus/p-progress/issues/10#issuecomment-795016415

Based on https://github.com/sindresorhus/p-all/pull/9

### CODE REVIEW

**Comment by Richienb** [comment]
The build is failing because the newest version of `tsd` expects Node.js 10: https://github.com/Richienb/p-progress/runs/2718053643?check_suite_focus=true#step:5:20

After this PR is merged, I'll do https://github.com/sindresorhus/p-progress/issues/10 and https://github.com/sindresorhus/p-progress/issues/18 and also require Node.js 12 before the next version is released.

**Comment by sindresorhus** [comment]
@privatenumber Would you be willing to help review?

**Review by sindresorhus — APPROVED** [review]

File: index.d.ts [hunk 1]
```diff
@@ -335,394 +104,16 @@ declare class PProgress<ValueType> extends Promise<ValueType> {
...
-	static all<AllValuesType>(
-		promises: Iterable<
-			PromiseLike<AllValuesType> | PProgress.PromiseFactory<AllValuesType>
-		>
-	): PProgress<Iterable<AllValuesType>>;
+	static all<Promises extends Array<PProgress.PromiseFactory<unknown> | PromiseLike<unknown>>>(
+		promises: readonly [...Promises],
+		options?: PProgress.Options
+	): PProgress<{
+		[Promise_ in keyof Promises]: Promises[Promise_] extends PromiseLike<unknown> ? Awaited<Promises[Promise_]> : (Promises[Promise_] extends PProgress.PromiseFactory<unknown> ? Awaited<ReturnType<Promises[Promise_]>> : Promises[Promise_])
```
Comment (privatenumber) [inline index.d.ts]: Personally find this hard to read. I don't think the TS plugin for ESLint has a rule for this but, I might suggest something like this:

```suggestion
		[Promise_ in keyof Promises]: (
			Promises[Promise_] extends PromiseLike<unknown>
				? Awaited<Promises[Promise_]>
				: (
					Promises[Promise_] extends PProgress.PromiseFactory<unknown>
						? Awaited<ReturnType<Promises[Promise_]>>
						: Promises[Promise_]
				)
		)

```

**Review by privatenumber — APPROVED** [review]
Looks good! Glad it was a simpler change than I expected.

Seems like the other Iterators (eg. Set) you had concerns for aren't an issue either 👍

It should also be noted that this is a breaking change in the type API

**Pull request merged**